quark/subproject.py

```diff
@@ -978,10 +983,10 @@ def _extract(self, archive_path: str, tempdir: str):
         extract_dir = join(tempdir, "extract")
         os.mkdir(extract_dir)
 
-        if archive_path.endswith((".tar.bz2", ".tar.gz", ".tar.xz")):
+        if archive_path.endswith((".tar.bz2", ".tar.gz", ".tar.xz")) or self.extract_format == "tar":
             with tarfile.open(archive_path, "r") as tar:
                 tar.extractall(extract_dir)
-        elif archive_path.endswith(".zip"):
+        elif archive_path.endswith(".zip") or self.extract_format == "zip":
             with zipfile.ZipFile(archive_path, "r") as zip:
                 zip.extractall(extract_dir)
```

Comment:
This means that e.g. a `.tar.bz2` will be extracted as a tar even if the user specified `extract=zip`, which is bizarre thing to do, but doing differently is surprising. This should be handled in a more predictable way:

```
if extract_format == 'true':
    if archive_path.endswith((".tar.bz2", ".tar.gz", ".tar.xz")):
        extract_format = "tar"
    elif archive_path.endswith(".zip"):
        extract_format = "zip"
    else:
        extract_format = "unknown"

if extract_format == "tar":
    with tarfile.open(archive_path, "r") as tar:
        tar.extractall(extract_dir)
elif extract_format == "zip":
    with zipfile.ZipFile(archive_path, "r") as zip:
        zip.extractall(extract_dir)
else:
    raise QuarkError("Unsupported format %s (%s)" % (extract_format, archive_path))
```

or something to this extent.

Reply:
Correct, the predictable check should first follow the format and later autodetect:

```python
self.extract_format == "zip"  or archive_path.endswith(".zip"):
```

Reply:
Nope! This gives the wrong result if you ask for tar extraction on a zip file. It's not just a matter of order, the point is that you shouldn't ever look at the extension isn't `true`, not just if it isn't the one you are considering at the moment. I find thus easiest to understand with the two-phase code above (reducing the "auto" case to the explicit one if necessary, and then applying the explicit logic), but you can also write it "inline" like
```
if self.extract_format == "zip" or (self.extract_format == "true" and archive_path.endswith(".zip"):
```